### DISCUSSION TITLE
Keyball61 firmware spec

### DISCUSSION THREAD

**Discussion by koron** [Yowkees/keyball#79] [General]
This discussion is for the specification and changes for Keyball61 firmware.
(Japanese) このディスカッションはKeyball61のファームウェアの仕様と、それに向けての既存のコードの変更について調整するためのものです。

And this includes multiple topics. Each comments may have different topics, please add reply to comment you interested.
(Japanese) ココには複数のトピックが含まれています。コメントごとに別のトピックを取り扱うので、気になったトピックにはreplyを書いてください。

---

[Latest code of Keyball61 is at here](https://github.com/koron/qmk_firmware/tree/develop-keyball61/keyboards/keyball/keyball61)

**Comment by koron** [comment]
Keyball61とKeyball46でソースコードを共用するために、ディレクトリ構成を以下のように変えるつもりです。

* 名前変更 keyball/rev1 → keyball/keyball46
* 移動 keyball/keymaps → keyball/keyball46/keymaps
* keyball 直下の雑多なファイルを keyball/keyball46にマージ

これにより光学センサー周りのコードをKeyball46とKeyball61で共用できるようになります。
なおビルドコマンドはそれぞれ以下のようになる予定です。

```console
# Keyball46
$ make keyball/keyball46/ball:default
$ make keyball/keyball46/noball:default

# Keyball61
$ make keyball/keyball61:test
```

@Yowkees 懸念点等ありましたらお知らせください。

**Reply by Yowkees** [reply]
分かりやすく効率的で良いと思います！
ありがとうございますm(_ _)m

**Comment by koron** [comment]
Keyball61 の追加に伴い既存のKeyball46の `PRODUCT` を現在の `Keyball` から `Keyball46` に変更したいです。
以下の部分です。
https://github.com/Yowkees/keyball/blob/92516b80911e3b0abe445482f451612ce9051db0/qmk_firmware/keyboards/keyball/rev1/config.h#L27

これによりPCに接続した際に表示される名称が変わります。
加えてもしかしたらRemapへの登録名も変更する必要があるかもしれません。

なお当然ですがKeyball61のほうは `Keyball61` とする予定です。

それにより接続時に何を繋げたかがわかりやすくなります。

**Reply by Yowkees** [reply]
名称変更了解です！
Remapの登録内容確認してみます。

**Comment by koron** [comment]
決めなきゃいけないこと:

* [x] `PRODUCT_ID` USBのPIDに使う。今は 0x0100 を仮置きとして使用している
* [x] デフォルトキーマップ
* [x] VIA(Remap)用の定義JSON

悩んでること:

* 構成に応じた(右ボール or 左ボール or 両ボール) キーマップを自動で選択する方法がないか

    Keyball46ではファームウェアを分けていたがスマートとはいえない。
    最大にイケてないポイントはVIAのJSONを4種類定義してPRODUCT_IDを4つ消費していた点。
    いっそVIAではボール有側で使えない3つのキーを注釈付き(≒色を変える)で表示してしまうというのもありか?
    もしくはVIAのLAYOUTで表現できるかもしれない。

**Reply by koron** [reply]
RemapでのLayout Optionsの参考動画
* セレクトボックス版: https://twitter.com/kaoriya/status/1483287530342862849
* トグル2つ版: https://twitter.com/kaoriya/status/1483289307993767940

**Reply by koron** [reply]
Layout Optionsはセレクトボックス版に決まった。

**Comment by koron** [comment]
pre-definedのキーマップを1つにしたい、という提案です。

Keyball46の時はright, left, dual ballについてpre-definedキーマップを提供していました。
これはKeyball46の左右の配列がドラスティックに異なっていたためです。

しかしKeyball61は配列自体は全く同じで、ボールの有無で3つのキーが無効化/有効化されるだけです。
これならばrightをベースに右ボールのある場所に本来あるはずだった3つのキーに適当なコードを割り当てておいて
「右で使うことを念頭においてるけど、かろうじて左やダブルボールでも使えなくはないかもね」
という意味でただ1つの **universal** キーマップを提供するのがスマートに思えます。

これによりファームウェアのバリエーションが格段に減り test, default, via の3種類だけにすると
準備するほうも楽ですが、何よりもユーザーにとってもわかりやすくて良いなという考えです。

**Reply by Yowkees** [reply]
ご尽力本当にありがとうございます！！
ご提案通り、ぜひ１つにまとめたいと思います🙏
現在ブランクとなっている右手の３キーの割当、パッとは浮かばないのですが、何か良いアイデアありますでしょうか？

**Reply by koron** [reply]
左手ボールの場合は左の3キー`LALT, LGUI, LT(1, LANG2)`がつぶれてしまいます。
で右手側にはRALTがありますので、必須と言えるのは `RGUI, LT(1, LANG2)` かなと考え
[現状のファームウェア](https://github.com/koron/qmk_firmware/blob/50c869f449862b68245e7febd2b80dc923e09eb1/keyboards/keyball/keyball61/default_keymap.h#L27)では以下のように定義しています。

```
LT(1,KC_LANG2), KC_RGUI, _______
```
 
1つ埋まっていませんが、まぁこれはいい案を思いつくまでそのままでも良いだろうと考えています。

**Reply by Yowkees** [reply]
なるほど、そうですね。左手ボール仕様で欠損するキーにしておけばよさそうです！
ありがとうございます🙏

**Comment by koron** [comment]
OLEDの変更点:

Keyball46の時はOLEDの表示は Ball, Key の順序でしたが、Keyball61からはKey, Ballの順序にしたいです。
(いずれKeyball46の方もあわせて修正したいと考えています)

というのも Key, Ball の順序にしとけばキーボードの名称と同じになって一貫性があるなと。
最初はトラックボールの座標を見ることしか考えておらずあの順だったんですが、
よくよく考えたらKeyballなんだからKey, Ballだろうと。

**Reply by Yowkees** [reply]
Keyballのブランディングについてまで考えていただき、ホントに感謝です！
ぜひkey→ballの順序にしたいと思います🙏
